File: geobus-backend/src/main/java/com/geobus/marrakech/repository/StopRepository.java
package com.geobus.marrakech.repository;

import java.util.List;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import com.geobus.marrakech.model.Stop;

/**
 * Repository pour accéder aux données des stations de bus
 */
@Repository
public interface StopRepository extends JpaRepository<Stop, Long> {

    /**
     * Trouve toutes les stations dans une ville spécifique
     *
     * @param ville Nom de la ville
     * @return Liste des stations dans cette ville
     */
    List<Stop> findByVilleIgnoreCase(String ville);

    /**
     * Trouve la station la plus proche des coordonnées données
     * Utilise une requête native SQL avec le calcul de Haversine
     *
     * @param lat Latitude de l'utilisateur
     * @param lon Longitude de l'utilisateur
     * @return La station la plus proche
     */
    @Query(value = "SELECT * FROM stops ORDER BY " +
            "(6371 * acos(cos(radians(?1)) * cos(radians(latitude)) * cos(radians(longitude) - radians(?2)) + sin(radians(?1)) * sin(radians(latitude)))) ASC LIMIT 1",
            nativeQuery = true)
    Stop findNearestStop(Double lat, Double lon);

    /**
     * Trouve les stations dans un rayon donné
     */
    @Query("SELECT s FROM Stop s WHERE " +
            "(6371 * acos(cos(radians(:lat)) * cos(radians(s.latitude)) * " +
            "cos(radians(s.longitude) - radians(:lng)) + " +
            "sin(radians(:lat)) * sin(radians(s.latitude)))) < :radius")
    List<Stop> findStopsWithinRadius(@Param("lat") Double latitude,
                                     @Param("lng") Double longitude,
                                     @Param("radius") Double radiusKm);

    /**
     * Recherche par nom de station
     */
    List<Stop> findByStopNameContainingIgnoreCase(String name);
}

File: geobusandroid/app/src/main/java/com/geobus/marrakech/repository/StopRepository.kt
package com.geobus.marrakech.repository

import com.geobus.marrakech.api.ApiClient
import com.geobus.marrakech.model.Stop
import com.geobus.marrakech.model.TimeEstimation
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Repository pour les opérations liées aux stations de bus
 */
class StopRepository {

    private val stopService = ApiClient.stopService
    private val apiService = ApiClient.getApiService()

    /**
     * Récupère toutes les stations de bus à Marrakech
     * @return Liste des stations ou null en cas d'erreur
     */
    suspend fun getAllStopsInMarrakech(): List<Stop>? = withContext(Dispatchers.IO) {
        try {
            val response = stopService.getAllStopsInMarrakech()
            if (response.isSuccessful) {
                response.body()
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Récupère une station par son ID
     * @return La station ou null en cas d'erreur
     */
    suspend fun getStopById(stopId: Long): Stop? = withContext(Dispatchers.IO) {
        try {
            val response = apiService.getStopById(stopId)
            if (response.isSuccessful) {
                response.body()
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Trouve la station la plus proche des coordonnées fournies
     * @return La station la plus proche ou null en cas d'erreur
     */
    suspend fun getNearestStop(lat: Double, lon: Double): Stop? = withContext(Dispatchers.IO) {
        try {
            val response = stopService.getNearestStop(lat, lon)
            if (response.isSuccessful) {
                response.body()
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Calcule le temps estimé pour atteindre une station spécifique
     * @return Informations sur le temps de trajet ou null en cas d'erreur
     */
    suspend fun getTimeToStop(userLat: Double, userLon: Double, stopId: Long): TimeEstimation? =
        withContext(Dispatchers.IO) {
            try {
                val response = stopService.getTimeToStop(userLat, userLon, stopId)
                if (response.isSuccessful) {
                    response.body()
                } else {
                    null
                }
            } catch (e: Exception) {
                null
            }
        }
}
